### PR TITLE
Topic/operational frames

### PR DESCRIPTION
**Pull request by fvalenza**
Reworked version of operational_frames :

### CODE REVIEW

File: src/multibody/model.hxx [hunk 1]
```diff
@@ -173,6 +164,77 @@ namespace se3
...
+  {
+    return std::find_if( operational_frames.begin(), operational_frames.end(), boost::bind(&Frame::name, _1) == name) != operational_frames.end();
+  }
+
+  inline const std::string & Model::getFrameName ( Index index ) const
+  {
+    return operational_frames[index].name;
+  }
+
+  inline const Model::Index& Model::getFrameParent( const std::string & name ) const
```
Comment (jcarpent) [inline src/multibody/model.hxx]: This method can lead to runtime error if name is not in the list.

File: src/multibody/model.hxx [hunk 2]
```diff
@@ -173,6 +164,77 @@ namespace se3
...
+  }
+
+  inline const SE3 & Model::getJointToFrameTransform( const std::string & name) const
+  {
+    assert(existFrame(name) && "The Frame you requested does not exist");
+    std::vector<Frame>::const_iterator it = std::find_if( operational_frames.begin()
+                                                        , operational_frames.end()
+                                                        , boost::bind(&Frame::name, _1) == name
+                                                        );
+    return it->frame_placement;
```
Comment (jcarpent) [inline src/multibody/model.hxx]: Same here in release mode.

File: src/python/frame.hpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+          .add_property("parent_id", &FramePythonVisitor::getParentId, &FramePythonVisitor::setParentId)
+          .add_property("placement", &FramePythonVisitor::getPlacementWrtParentJoint, &FramePythonVisitor::getPlacementWrtParentJoint)
+          ;
+      }
+
+
+      static std::string getName( const Frame & self) { return self.name; }
+      static void setName(Frame & self, const std::string & name) { self.name = name; }
+      static Index getParentId( const Frame & self) { return self.parent_id; }
+      static void setParentId(Frame & self, Index & parent_id) { self.parent_id = parent_id; }
```
Comment (jcarpent) [inline src/python/frame.hpp]: const and not a reference for the second argument

File: src/python/frame.hpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      template<class PyClass>
+      void visit(PyClass& cl) const 
+      {
+        cl
+          .def(bp::init<const std::string&,Index, const SE3_fx&> ((bp::arg("name"),bp::arg("parent id"), bp::arg("placement")),
+                "Initialize from name, parent id and placement wrt parent joint."))
+
+          .add_property("name", &FramePythonVisitor::getName, &FramePythonVisitor::setName)
+          .add_property("parent_id", &FramePythonVisitor::getParentId, &FramePythonVisitor::setParentId)
+          .add_property("placement", &FramePythonVisitor::getPlacementWrtParentJoint, &FramePythonVisitor::getPlacementWrtParentJoint)
```
Comment (jcarpent) [inline src/python/frame.hpp]: Do not use placement but rather framePlacement like jointPlacement..

File: src/python/model.hpp [hunk 5]
```diff
@@ -132,6 +132,10 @@ namespace se3
       .add_property("fix_hasVisual", bp::make_function(&ModelPythonVisitor::fix_hasVisual, bp::return_internal_reference<>())  )
       .add_property("fix_bodyNames", bp::make_function(&ModelPythonVisitor::fix_bodyNames, bp::return_internal_reference<>())  )
 
+      .def("getFrameParentJoint", &ModelPythonVisitor::getFrameParent)
+      .def("getFramePositionInParent", &ModelPythonVisitor::getJointToFrameTransform)
```
Comment (jcarpent) [inline src/python/model.hpp]: a good name here may just be getFramePlacement (it is implicit that it is regarding to the parent joint)

File: src/python/model.hpp [hunk 6]
```diff
@@ -132,6 +132,10 @@ namespace se3
       .add_property("fix_hasVisual", bp::make_function(&ModelPythonVisitor::fix_hasVisual, bp::return_internal_reference<>())  )
       .add_property("fix_bodyNames", bp::make_function(&ModelPythonVisitor::fix_bodyNames, bp::return_internal_reference<>())  )
 
+      .def("getFrameParentJoint", &ModelPythonVisitor::getFrameParent)
```
Comment (jcarpent) [inline src/python/model.hpp]: same here with getFrameParent

File: src/algorithm/operational-frames.hpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+   * @param[in]  model  The kinematic model
+   * @param      data   Data associated to model
+   * @warning    One of the algorithms forwardKinematics should have been called first
+   */
+  inline void framesForwardKinematic(const Model & model,
+                                          Data & data
+                                          );
+
+  /**
+   * @brief      Compute Kinematics of full model, then the position of each extra frame
```
Comment (jcarpent) [inline src/algorithm/operational-frames.hpp]: The doc needs an update regarding to your naming

**Comment by fvalenza** [comment]
@jcarpent new version. Is the computation of the jacobian ok ?

**Comment by jcarpent** [comment]
:+1:

**Pull request merged**